lib/logproto/logproto-framed-server.c

```diff
@@ -152,120 +156,178 @@ log_proto_framed_server_extract_frame_length(LogProtoFramedServer *self, gboolea
   return TRUE;
 }
 
+static void
+_adjust_buffer_if_needed(LogProtoFramedServer *self, const gsize minimum_space_required)
+{
+  /* We need at least minimum_space_required amount of free space in the buffer.
+   * Since we already moving memory, move it to the beginning. */
+  if (self->buffer_size - self->buffer_pos < minimum_space_required)
+    {
+      memmove(self->buffer, &self->buffer[self->buffer_pos], self->buffer_end - self->buffer_pos);
+      self->buffer_end = self->buffer_end - self->buffer_pos;
+      self->buffer_pos = 0;
+    }
+}
+
+/* Returns TRUE if successfully finished consuming the data. Othwerwise it is not finished, but
+ * there is nothing left to read (or there was a read error) and expects to be called again. */
+static gboolean
+_consume_trimmed_part(LogProtoFramedServer *self, gboolean *may_read,
+                      LogTransportAuxData *aux, LogProtoStatus *status)
+{
+  gboolean buffer_was_full;
+  /* Since trimming requires a full (buffer sized) message, the consuming
+   * always starts at the beginning of the buffer, with a new read. */
+  self->half_message_in_buffer = FALSE;
+
+  while (1)
+    {
+      if (!log_proto_framed_server_fetch_data(self, may_read, aux, status))
+        return FALSE;
+
+      // We have some data, but it is not enough.
+      if (self->buffer_end < self->frame_len)
+        {
+          self->frame_len -= self->buffer_end;
+          buffer_was_full = self->buffer_end == self->buffer_size;
+          self->buffer_end = 0;
+
+          // We might have more data, waiting for read.
```

Comment:
Correct me if I'm wrong, but this function can be reduced a bit:

```C
static inline gboolean
_is_trimmed_part_completely_fetched(LogProtoFramedServer *self)
{
  return self->frame_len <= self->buffer_end;
}

static gboolean
_consume_trimmed_part(LogProtoFramedServer *self, gboolean *may_read,
                      LogTransportAuxData *aux, LogProtoStatus *status)
{
  /* Since trimming requires a full (buffer sized) message, the consuming
   * always starts at the beginning of the buffer, with a new read. */
  self->half_message_in_buffer = FALSE;

  while (1)
    {
      if (!log_proto_framed_server_fetch_data(self, may_read, aux, status))
        return FALSE;

      if (_is_trimmed_part_completely_fetched(self))
        {
          self->buffer_pos += self->frame_len;
          return TRUE;
        }
      
      self->frame_len -= self->buffer_end;
      self->buffer_end = 0;
    }
}
```

Reply:
I would like to start a conversation on this topic.

If I remove the `buffer_was_full` logic, than the code will be reduced a bit. BUT I also have to remove the `return FALSE` statement at the end with it. (Which is meant to avoid an unnecessary `fetch` call.)
Eventually the code will be correct. We will return with FALSE, when we got EAGAIN in the next fetch attempt.

Reply:
The fact that the buffer was not full does not mean `fetch_data()` will not return anything in the next call. For example, when reading a TCP stream , `read()` can return data in any schedule, `EAGAIN` is the only guarantee that it can't read more in this round.

Reply:
Thak you! Fixed